pkg/services/ngalert/notifier/images.go

```diff
@@ -44,3 +51,101 @@ func (i imageStore) GetImage(ctx context.Context, uri string) (*images.Image, er
 		CreatedAt: image.CreatedAt,
 	}, nil
 }
+
+func (i imageProvider) GetImageURL(ctx context.Context, alert *alertingNotify.Alert) (string, error) {
+	uri, err := getImageURI(alert)
+	if err != nil {
+		return "", err
+	}
+
+	// If the identifier is a URL, validate that it corresponds to a stored, non-expired image.
+	if strings.HasPrefix(uri, "http") {
+		exists, err := i.store.URLExists(ctx, uri)
+		if err != nil {
+			return "", err
+		}
+		if !exists {
+			return "", images.ErrImageNotFound
+		}
+		return uri, nil
+	}
+
+	// If the identifier is a token, remove the prefix, get the image and return the URL.
+	token := strings.TrimPrefix(uri, "token://")
+	return i.getImageURLFromToken(ctx, token)
+}
+
+// getImageURLFromToken takes a token and returns the URL of the image that token belongs to.
+func (i imageProvider) getImageURLFromToken(ctx context.Context, token string) (string, error) {
+	image, err := i.store.GetImage(ctx, token)
+	if err != nil {
+		if errors.Is(err, models.ErrImageNotFound) {
+			return "", images.ErrImageNotFound
+		}
+		return "", err
+	}
+
+	if !image.HasURL() {
+		return "", images.ErrImagesNoURL
+	}
+	return image.URL, nil
+}
+
+func (i imageProvider) GetRawImage(ctx context.Context, alert *alertingNotify.Alert) (io.ReadCloser, string, error) {
+	uri, err := getImageURI(alert)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Check whether the uri is a URL or a token to know how to query the DB.
+	var image *models.Image
+	if strings.HasPrefix(uri, "http") {
+		image, err = i.store.GetImageByURL(ctx, uri)
+	} else {
+		token := strings.TrimPrefix(uri, "token://")
+		image, err = i.store.GetImage(ctx, token)
+	}
+	if err != nil {
+		if errors.Is(err, models.ErrImageNotFound) {
+			return nil, "", images.ErrImageNotFound
+		}
+		return nil, "", err
+	}
```

Comment:
wdyt about a tiny refactor: create a private method `getImage(ctx, alert) (image,error)` implement reading annotation and database logic in it, and then update GetRawImage and GetImageURL to use the result to do what needs to be done.

Reply:
I wouldn't do that for two main reasons:
1. There's no need to get the entire image when we want a URL and the URI is already a URL.
2. There are already too many jumps in code, and adding more helper functions would make it even worse.